src/python/pants/core/goals/package.py

```diff
@@ -178,5 +185,35 @@ async def package_asset(workspace: Workspace, dist_dir: DistDir) -> Package:
     return Package(exit_code=0)
 
 
+@dataclass(frozen=True)
+class TraverseIfNotPackageTarget(ShouldTraverseDepsPredicate):
+    package_field_set_types: FrozenOrderedSet[PackageFieldSet]
+    roots: FrozenOrderedSet[Address]
+    always_traverse_roots: bool = True  # traverse roots even if they are package targets
+
+    def __init__(
+        self,
+        *,
+        union_membership: UnionMembership,
+        roots: Iterable[Address],
+        always_traverse_roots: bool = True,
+    ) -> None:
+        object.__setattr__(self, "package_field_set_types", union_membership.get(PackageFieldSet))
+        object.__setattr__(self, "roots", FrozenOrderedSet(roots))
+        object.__setattr__(self, "always_traverse_roots", always_traverse_roots)
+        super().__init__()
```

Comment:
Is this because you need `__post_init__` called? Oof that's a footgun for sure. We might consider instrumenting a pytest autouse fixture which validates this is being called by any subclass?

I guess if you _dont_ define it though, what happens? Ideally it goes :boom: in `__hash__`?

Reply:
If you don't, MyPy complains because not all of the dataclass attributes get set.

Reply:
:chefs kiss: